### PR TITLE
APPENG-98: added the config for fan-out topology

### PR DESCRIPTION
**Pull request by sgahlot**
The routers are each running in a separate VM and the two live brokers are also running in separate VMs.

### CODE REVIEW

**Review by mpaulgreen — COMMENTED** [review]
@sgahlot lets have a meeting on these changes first.

**Review by mpaulgreen — COMMENTED** [review]
@sgahlot don't you need connectors for router1 so that the connection between router2 and router3 is bidirectional?

**Comment by sgahlot** [comment]
@mpaulgreen for which router are you referring to when you mentioned we need connector? There is a connector defined in router2 and router3 to point to router1.

Also, what do you mean by bidirectional between router2 and router3? There is no direct connection between router2 and router3, but once they're part of the topology, they are all connected.

**Comment by mpaulgreen** [comment]
- The following comments. are for only connectors and listener configuration.
@sgahlot as per the topology diagram the router1. 
     - Listening to client connection (Missing)
     - Listening to other routers (Present)
     - Since it is the hub where it is fanning out to Router2 and Router3 - It needs a connector to Router2 and Router3
    Router2 - 
       - Listening to other routers (Missing)
       - Connector to router1 (Not Required as we moved it to router1)
       - Connector to broker (present)
      Router3 - Same as Router2

And now take a moment to think, we are creating a pattern for hub accepting client connections and creating a connection to other routers and router2 and router3 connecting to brokers. We can now scale this pattern horizontally for both hub router and router connecting to broker.

Let me know your thoughts. You may have your code/configuration working, but can we improve on this?

Thanks
Mriganka

**Comment by sgahlot** [comment]
Hi Mriganka,

I think it will be better if we have the meeting that you asked to have
first. But, I'll to answer your questions here.

   1. Listening to the client connection:
      - The listener element takes care of that (default amqp listener)
   2. Listening to other router - present as you mentioned
   3. Fanning out:
      - Listening to other router:
         - The listener with port 5773 takes care of that (in router1
         config)
         - Connector to router 2 and router 3:
         - Don't need it as the listener is enough. If you add the
         connectors to both of these routers, it's same as adding a
listener, but
         you're limiting your scalability as you'll now need to add
the connector
         for each new router you want to connect to the hub. With the
listener in
         the hub, you don't need that.
         - Also, the connectivity is same whether you add the listener or
         add the connectors in the hub as you can see in the logs if
you run all 3
         routers.
      - Listening to other routers:
      - Which config are you referring to here? Router1 is listening to all
         the routers that will try to connect to it. There is nothing
in your design
         that specifies the connection between router 2 and 3 - if
you/re referring
         to that.

With the router1 accepting connections from clients (producers/consumers)
as well as other routers, it's exactly what the design asked for.
And, router2 and router3 are connecting to the brokers as well as to the
hub (router1), so I'm not clear on what's missing here and what can be
improved.

If you're still not clear on the above then it's better to have a meeting
where I can show you what I mean.

Thanks.
Sandip.

File: manualconfig/routertopology/fanout/readme.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+* Add the config given below at the end of the config file
+
+
+<details>
+  <summary>Click to expand or collapse the config</summary>
+
+```shell
+# Listener for inter-router comm
+listener {
```
Comment (sgahlot) [inline manualconfig/routertopology/fanout/readme.md:44]: Listener for client connections - for accepting connections from producer/consumers

File: manualconfig/routertopology/fanout/readme.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+* Add the config given below at the end of the config file
+
+<details>
+  <summary>Click to expand or collapse the config</summary>
+
+```shell
+
+# Connection to the hub
+connector {
```
Comment (sgahlot) [inline manualconfig/routertopology/fanout/readme.md:82]: Connection to the hub (router1)

File: manualconfig/routertopology/fanout/readme.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+* Add the config given below at the end of the config file
+
+<details>
+  <summary>Click to expand or collapse</summary>
+
+```shell
+
+# Connection to the hub
+connector {
```
Comment (sgahlot) [inline manualconfig/routertopology/fanout/readme.md:145]: Connection to the hub (router1)

**Comment by sgahlot** [comment]
The `listener` element that accepts connection from the clients (producer/consumer) is missing from the readme as that's the default element in the router config file and doesn't need to be changed as part of this topology. Hence, it is not provided in the readme.

**Comment by mpaulgreen** [comment]
@sgahlot Thanks for your reply. You have explained your thoughts very well. I agree with you having connectors in hub router, we will need additional connectors for every router joining in. By the way topology diagram is at architectural level and not considered as a design. Now when we start implementing in cloud, here are some of the best practices.
```
Generally, inter-router connections should always be established from more private to more public. For example, to connect a router in a private IP network to another router in a public location (such as a public cloud provider), the router in the private network must be the "connector" and the router in the public location must be the "listener". This is because the public location cannot reach the private location by TCP/IP without the use of VPNs or other firewall features designed to allow public-to-private access.
```

**Pull request merged**